### PR TITLE
Binplace package file list

### PR DESCRIPTION
**Pull request by ericstj**
/cc @weshaggard @chcosta @joperezr 

This switches binplacing for packaging to use file lists rather than copies.

### CODE REVIEW

File: binplace.targets [hunk 1]
```diff
@@ -88,17 +87,35 @@
...
+        <TargetPath Condition="$([System.String]::Copy('%(FullPath)').StartsWith('$(ProjectDir)'))">src/$([System.String]::Copy('%(FullPath)').Substring($(_projectDirLength)).Replace('\', '/'))</TargetPath>
+        <IsSourceCodeFile>true</IsSourceCodeFile>
+      </_itemsToSave>
+    </ItemGroup>
+    
+    <Message Importance="low" Text="PackageFileDir: @(PackageFileDir)" />
+
+    <SaveItems ItemName="FileToPackage"
+               Items="@(_itemsToSave)"
+               Files="@(PackageFileDir->'%(Identity)\$(_propsFilename).props')" />
```
Comment (joperezr) [inline binplace.targets]: Will these new lists or props files be packaged as well? Or will we just use them to know what goes in each package?

Comment (ericstj) [inline binplace.targets]: Not packaged.  They just get imported by the pkgrpojs to create the file items (rather than a wildcard in the pkgproj).

File: pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj [hunk 2]
```diff
@@ -42,30 +42,32 @@
...
+  <Import Condition="'$(PackageTargetRuntime)' == ''" Project="$(RefBinDir)\*.props" />
   <ItemGroup Condition="'$(PackageTargetRuntime)' == ''">
-    <ProjectReference Include="@(Project)" />
-
     <!-- Include refs -->
-    <RefFile Include="$(RefBinDir)/*.*">
-      <TargetPath>ref/$(TargetFramework)</TargetPath>
-    </RefFile>
-    <File Include="@(RefFile)" />
+    <File Include="@(FileToPackage)">
```
Comment (chcosta) [inline pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj]: are you planning to move any of this to a shared file?

File: src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj [hunk 3]
```diff
@@ -7,7 +7,6 @@
     <DocumentationFile>$(OutputPath)$(MSBuildProjectName).xml</DocumentationFile>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <IsPartialFacadeAssembly Condition="'$(TargetGroup)'=='netfx'">true</IsPartialFacadeAssembly>
-    <PackageTargetFramework Condition="'$(TargetGroup)' == 'netcoreapp'">netcoreapp;portable-net45+win8+wp8+wpa81</PackageTargetFramework>
```
Comment (weshaggard) [inline src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj]: Why was this needed for this change?

File: binplace.targets [hunk 4]
```diff
@@ -68,17 +65,19 @@
     </BinplaceConfiguration>
   </ItemGroup>
 
-  <Target Name="BinPlaceStuff" DependsOnTargets="$(BinPlaceStuffDependsOn)" AfterTargets="CopyFilesToOutputDirectory">
-    <Message Importance="low" Text="IsRuntimeAssembly: $(IsRuntimeAssembly)" />
-    <Message Importance="low" Text="IsReferenceAssembly: $(IsReferenceAssembly)" />
-    <Message Importance="low" Text="IsRuntimeAndReferenceAssembly: $(IsRuntimeAndReferenceAssembly)" />
-  </Target>
+  <Target Name="BinPlace"
+          DependsOnTargets="GetBinplaceDirs;BinPlaceFiles;BinplaceProps"
```
Comment (weshaggard) [inline binplace.targets]: nit: consistency some Binplaces have capital P and some have lowercase.

File: binplace.targets [hunk 5]
```diff
@@ -68,17 +65,19 @@
...
+          AfterTargets="CopyFilesToOutputDirectory"
+          Condition="'$(BinPlaceRef)' == 'true' OR '$(BinPlaceRuntime)' == 'true'" />
```
Comment (weshaggard) [inline binplace.targets]: Is this condition true for shims?

Comment (weshaggard) [inline binplace.targets]: We might want to move this conditions to GetBinplaceItems.

Comment (ericstj) [inline binplace.targets]: Yes.

File: src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj (hunk 3 above)
Comment (ericstj) [inline src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj]: I needed it for an earlier version of the change where I was using GetFilesToPackage.  I noticed that it was using an invalid TFM (netcoreapp with no version).  Moreover the property isn't needed at all because the library doesn't have a pkgproj.

File: binplace.targets (hunk 4 above)
Comment (ericstj) [inline binplace.targets]: Ok, I'll clean that up.

File: pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj (hunk 2 above)
Comment (ericstj) [inline pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj]: I should eventually.  We already have lots of duplication here so I'm not tackling that just yet.

File: binplace.targets [hunk 6]
```diff
@@ -88,17 +87,35 @@
...
-          DestinationFolder="%(RuntimeDir.Identity)"
-          SkipUnchangedFiles="true"
-          OverwriteReadOnlyFiles="true"
-          Retries="$(CopyRetryCount)"
-          RetryDelayMilliseconds="$(CopyRetryDelayMilliseconds)"
-          UseHardlinksIfPossible="$(BinplaceUseHardlinksIfPossible)">
-      <Output TaskParameter="DestinationFiles" ItemName="FileWrites" />
-    </Copy>
+  <UsingTask TaskName="SaveItems" AssemblyFile="$(CoreFxToolsTaskDir)CoreFx.Tools.dll"/>
+  <Target Name="BinplaceProps"
```
Comment (weshaggard) [inline binplace.targets]: We should probably put this into the packagelib.targerts

Comment (ericstj) [inline binplace.targets]: Yeah.  We can do so when merging binplace.targets to buildtools.  I didn't want to fork packaging in corefx yet since I've needed to make cross-repo changes to both packaging targets and tasks recently.
